torchrl/data/replay_buffers/ray_buffer.py

```diff
@@ -0,0 +1,199 @@
+
+    The writer, sampler and storage should be passed as constructors to prevent serialization issues.
+    Transforms constructors should be passed through the `transform_factory` argument.
+
+    Example:
+        >>> import asyncio
+        >>> from tensordict.nn import TensorDictModule
+        >>> from torch import nn
+        >>> from torchrl.collectors.distributed.ray import RayCollector
+        >>> from torchrl.data.replay_buffers.ray_buffer import RayReplayBuffer
+        >>> from torchrl.envs.libs.gym import GymEnv
+        >>>
+        >>> async def main():
+        ...     # 1. Create environment factory
+        ...     def env_maker():
+        ...         return GymEnv("Pendulum-v1", device="cpu")
+        ...
+        ...     policy = TensorDictModule(
+        ...         nn.Linear(3, 1), in_keys=["observation"], out_keys=["action"]
+        ...     )
+        ...
+        ...     buffer = RayReplayBuffer()
+        ...
+        ...     # 2. Define distributed collector
+        ...     remote_config = {
+        ...         "num_cpus": 1,
+        ...         "num_gpus": 0,
+        ...         "memory": 5 * 1024**3,
+        ...         "object_store_memory": 2 * 1024**3,
+        ...     }
+        ...     distributed_collector = RayCollector(
+        ...         [env_maker],
+        ...         policy,
+        ...         total_frames=600,
+        ...         frames_per_batch=200,
+        ...         remote_configs=remote_config,
+        ...         replay_buffer=buffer,
+        ...     )
+        ...
+        ...     print("start")
+        ...     distributed_collector.start()
+        ...
+        ...     while True:
+        ...         while not len(buffer):
+        ...             print("waiting")
+        ...             await asyncio.sleep(1)  # Use asyncio.sleep instead of time.sleep
+        ...         print("sample", buffer.sample(32))
+        ...         # break at some point
+        ...         break
+        ...
+        ...     await distributed_collector.async_shutdown()
+        >>>
+        >>> if __name__ == "__main__":
+        ...     asyncio.run(main())
+
+    """
+
+    def __init__(self, *args, ray_init_kwargs=None, **kwargs) -> None:
+        if not _has_ray:
+            raise RuntimeError(
+                "ray library not found, unable to create a RayReplayBuffer. "
+            ) from RAY_ERR
+        if not ray.is_initialized():
+            if ray_init_kwargs is None:
+                from torchrl.collectors.distributed.ray import DEFAULT_RAY_INIT_CONFIG
+
+                ray_init_kwargs = DEFAULT_RAY_INIT_CONFIG
+            ray.init(**ray_init_kwargs)
+
+        remote_cls = ReplayBuffer.as_remote(DEFAULT_REMOTE_CLASS_CONFIG).remote
+        self._rb = remote_cls(*args, **kwargs)
+
+    def sample(self, *args, **kwargs):
+        pending_task = self._rb.sample.remote(*args, **kwargs)
+        return ray.get(pending_task)
```

Comment:
very n00b qn re how sharing this across distributed workers works

is the idea that by initializing RayReplayBuffer in __main__ and then passing it to each DistributedActorWorker, when .sample() in one of the DistributedActorWorker's methods is called it would do the .get on the resources that have been allocated to that specific DistributedActorWorker?

Reply:
Yes, not on main but within the worker

Reply:
@vmoens How would one use RayReplayBuffer in the context of PyTorch DDP/FSDP? Is this even possible? For example, can there be a single RayReplayBuffer that each DDP/FSDP process uses?

Reply:
Yes that should work: instantiate somewhere and pass the remote handler around.